hotplug-nics: Add pod interface naming change content (#217)
kubevirt/community: design-proposals/nic-hotplug/nic-hotplug.md

[the VM API examples section](#vms), while the proposed API changes for the VMI
object can be seen in [the VMI API examples section](#vmis).

### Pod interface naming

> As of today, the VMs’ pod interfaces names has an ordinal factor 
> (`net1`, `net2`, …, `netX`) which is derived from the specified in the VMI spec,
> the interfaces names are specified in the virt-launcher pod 
> `k8s.v1.cni.cncf.io/networks` annotation, and eventually the pod is created with 
> those interface names.

Reviewer: ```suggestion
The VMs’ pod interface names are ordinal based, derived from their order in the VMI spec;
they are requested from Multus by specifying any additional networks using the `k8s.v1.cni.cncf.io/networks` annotation on the launcher pod (which is templated by KubeVirt).
```

I guess this could still be improved, but, I also think it makes a decent job at getting the message across.

Given a VMI with 2 secondary interfaces its pod `k8s.v1.cni.cncf.io/networks` 
annotation would have `net1, net2, net3` elements.
When removing interface `net2`, the annotation will change to `net1`, `net3`, 
which makes associating the pod to VM interfaces impossible.

In order to enable unplug functionality virt-launcher pod network interfaces names 

> shall be changed in a way to won't have an ordinal factor.

Reviewer 2: ```suggestion
shall be changed in a way that they won't have an ordinal factor.
```

> In order to enable unplug functionality virt-launcher pod network interfaces names 
> shall be changed in a way to won't have an ordinal factor.

Reviewer: This sentence reads weird IMO. I think the subject if the name of the pod interfaces.

What do you think about:
```suggestion
Thus, the name of the `virt-launcher` pod network interfaces must be generated without relying on the interface ordinal, allowing the unplug feature.
```

The pod interface name will be derived from the `kubevirt-spec-iface-name`; 
we'll simply compute an Hash of the interface name
(which is guaranteed to be unique within each VMI), and ensure all generated

> names for the pod's networking infrastructure are accepted by the kernel.

Reviewer 2: Could you please add a note explaining the requirements for the name to be accepted by the kernel?

Refer to the following list for examples of names on pod networking infra:
- VM interface name: `iface1`
- pod interface name: `7e0055a6880`
- in-pod bridge name: `k6t-7e0055a6880`
- dummy pod nic name: `7e0055a6880-nic`

The proposed algorithm is SHA256; here's a minimal implementation:
```golang
func PodInterfaceName(vmiSpecInterfaceName string) string {
    // allows the dummy pod suffix (`-nic`) to fit the kernel limitation of 15 chars.
	const MaxIfaceNameLen = 11 
	hash := sha256.New()
	_, _ = io.WriteString(hash, vmiSpecInterfaceName)

> 	return fmt.Sprintf("%x", hash.Sum(nil))[:MaxIfaceNameLen]

Reviewer: nit: the spacing is a bit off here.

```yaml
spec:
  networks:
  - name: blue-network
    multus: ...
  - name: red-network
    multus: ...
  - name: green-network
    multus: ...
```

The pod's Multus networks annotation will look like so:
```json
"k8s.v1.cni.cncf.io/networks": [
  {"interface": "net1", ...},
  {"interface": "net2",  ...},  
  {"interface": "net3",  ...}
]
```

In the scenario where "red-network" interface is unplugged, the annotation will change  as follows:
```json
"k8s.v1.cni.cncf.io/networks": [
  {"interface": "net1", ...},
  {"interface": "net3",  ...}
]
```
Now it's impossible to associate between all VMI spec and pod interfaces.

Thus, the name of the `virt-launcher` pod network interfaces must be generated 
without relying on the interface order in the spec, allowing the unplug feature.

The pod interface name will be derived from the `kubevirt-spec-iface-name`; 
we'll simply compute a Hash of the interface name
(which is guaranteed to be unique within each VMI), and ensure all generated
names for the pod's networking infrastructure are accepted by the kernel.
Refer to the following list for examples of names on pod networking infra:
- VM interface name: `iface1`

> - pod interface name: `7e0055a6880`

Reviewer: this is not aligned w/ the implementation proposed in https://github.com/kubevirt/kubevirt/pull/9268#discussion_r1196207065

the `k8s.v1.cni.cncf.io/networks` annotation on the pod holding the VM, which
in turn causes multus to hotplug an interface into the pod.

> The request to plug/unplug will then be forwarded to the correct `virt-handler`.

Reviewer 2: How is the request forwarded to the correct `virt-handler`?

The aforementioned update will trigger multus to start the CNI ADD flow for the
network named `macvlan-conf-2`.

### Unplug for pods

> Following #hotplug-for-pods example, to unplug an interface, update the pod to:

Reviewer: I am assuming this should be a link instead.  Please fix it.

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: pod-case-03
  annotations:
    k8s.v1.cni.cncf.io/networks: macvlan-conf-2
spec:
  containers:
  - name: pod-case-03
    image: docker.io/centos/tools:latest
    command:
    - /sbin/init
```

The aforementioned update will trigger multus to start the CNI DEL flow for the
network named `macvlan-conf-1`.

## Backward Compatibility
### Legacy VM's virt-launcher pods interface naming
*Legacy VM - running VMs that were created prior to Kubevirt upgrade

> which run on top of previous version virt-launcher pod*

Reviewer 2: This sentence is a bit vague, could you please rephrase?

The proposed solution is having virt-controller to create the migration target pod with  
`k8s.v1.cni.cncf.io/networks` annotation with same pod interface names as the migration 
source pod annotation.

The migration target pod interfaces names will then match the names in the 
incoming migration domain XML and the migration process will start.

In case the user migrates the VM again, same as before, the migration target 
`pods k8s.v1.cni.cncf.io/networks` annotation value will have the same interface 
names as in the migration source annotation.

### Unplug interface of a legacy VM
#### Story 1
Running legacy VM with secondary networks, with the following networks in the spec:
```yaml
spec:
networks:
- name: blue-network
  multus:
  networkName: blue-net-br
- name: red-network
  multus:
  networkName: red-net-br
```
The VM pod network-status annotation will look as follows:
```yaml
k8s.v1.cni.cncf.io/networks-status: '[
    { "name": "blue-net-br", "interface": "net1", ...},
    { "name": "red-net-br", "interface": "net2", ...},
]'
```

Unplugging `net1` and `net2` should be blocked because it makes mapping between 

> the VMI networks names and the pod interfaces names impossible.

Reviewer 2: This sentence is written from Multus' perspective (an implementation detail).
Please consider changing the explanation to a VM's perspective.

```yaml
k8s.v1.cni.cncf.io/networks-status: '[
    { "name": "blue-net-br", "interface": "net1", ...},
    { "name": "red-net-br", "interface": "net2", ...},
]'
```

Unplugging `net1` and `net2` should be blocked because it makes mapping between 
the VMI networks names and the pod interfaces names impossible.

#### Story 2
Running legacy VM that was migrated following Kubevirt upgrade, and has new interfaces that were hot-plugged into it.
The VM pod will have interfaces with names in the form of the old name scheme, and some in the form of the new name scheme.

The VMI networks spec will be like so:
```yaml
spec:
networks:
- name: blue-network
  multus:
  networkName: blue-net-br
- name: red-network
  multus:
  networkName: red-net-br
- name: green-network
  multus:
  networkName: green-net-br
- name: yellow-network
  multus:
  networkName: yellow-net-br
```
The VM pod network-status annotation will look as follows:
```yaml
k8s.v1.cni.cncf.io/networks-status: '[
    { "name": "blue-net-br", "interface": "net1", ...},
    { "name": "red-net-br", "interface": "net2", ...},
    { "name": "green-net-br", "interface": "netXYZ123", ...},

>     { "name": "yellow-net-br", "interface": "netABC456", ...},

Reviewer: Ah of course. My bad.

The aforementioned update will trigger multus to start the CNI ADD flow for the
network named `macvlan-conf-2`.

### Unplug for pods

> Following [hotplug for pods example](#Hotplug-for-pods example), to unplug an interface, update the pod to:

Reviewer: this link is broken, please update it.
@ormergi

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: pod-case-03
  annotations:
    k8s.v1.cni.cncf.io/networks: macvlan-conf-2
spec:
  containers:
  - name: pod-case-03
    image: docker.io/centos/tools:latest
    command:
    - /sbin/init
```

The aforementioned update will trigger multus to start the CNI DEL flow for the
network named `macvlan-conf-1`.

## Backward Compatibility
### Legacy VM's virt-launcher pods interface naming
*Legacy VM - running VMs prior to KubeVirt version that introduce
the pod interface naming change.

> Such VMs runs on top of an old virt-launcher pod*

Reviewer: ```suggestion
These VMs run in an old virt-launcher pod*
```

All these tests have as pre-requirements that the `HotplugInterfaces` feature
gate is enabled, **and** a secondary network provisioned.

### `virt-launcher` pods network interfaces naming
* VMs that were created with secondary interfaces can be migrated by 
KubeVirt upgrade (when workload-strategy is set to Migrate), and after 
Kubevirt upgrade following user request.

It should cover changes around virt-launcher pod interface name change and in general cover 

> backward compatibility for changes related to the networking code.

Reviewer: I'm afraid I don't follow this section. Why is this in the middle of a section about tests ?

Author: This section describes how the pod interface nameing change should be tested, its a sub-section of `Functional Testing Approach`

Reviewer: `Functional Testing Approach` lists in its own heading the KubeVirt required tests (well ... an approximation of the tests).

Then it has a dedicated sub-heading about what tests must be made on the components that are not KubeVirt - i.e.  multus dynamic controller

I find it confusing to see a new sub-heading added pointing at an existing component. 

But honestly I don't have a better suggestion. It feels extra-weird since it feels you're adding design info for something that has already happened :) IMO, there's no point "designing" tests that already exist. 

Feel free to keep this if you want; IMO it's strange, but, I can live with it.

Author: I can remove the sub-heading, will that make sense to you?

Author: related to https://github.com/kubevirt/community/pull/217#discussion_r1183688449

> backward compatibility for changes related to the networking code.

Author: Where you suggest moving it?

> backward compatibility for changes related to the networking code.

Reviewer: You could omit the heading and re-structure the previous one to list all the tests we want to have in kubevirt.

But honestly, I would not add this; it feels like you're "designing" a test that IIUC already exists and was merged. 
Or am I missing something ? 

Also, feel free to wait for other opinions. I don't consider this a blocker of this PR.

3. Add a controller monitoring pod attachment updates
4. **C** Consume this dynamic networks functionality via CNAO
5. **K** Add the hot-plug functionality to KubeVirt for L2 and L3 networks
     (with IPAM enabled on the pod interface)`
6. **K** Change virt-launcher pod network interfaces name scheme
7. **K** Add a remove-interface command at `virtctl` and correspond endpoints at `virt-api`, **support VMI objects only**
8. **K** Detach the requested interface from the guest through Libvirt API, **support VMI objects only**.
9. **K** Extend the `InterfaceRequests` API to support remove-interface requests.
10. **K** Extend the remove-interface `virtctl`'s command and `virt-api`'s endpoints to support `VirtualMachine`.
11. **K** Implement `virt-controller` pod annotation patching for unplug requests.
12. **K** Cleanup the unplugged interface's bridge and tap-device from virt-launcher pods.
13. **K** Shut down the unplugged interface IPAM DHCP server instance.

**Notes:**
* the action items listed above have either `M`, `K`, or `C` to
indicate in which project should it be implemented.
* the MVP version would be composed of steps 1 through 4, inclusive.

> * the MVP for unplug functionality would be composed of steps 6 through 8, inclusive.

Author: We do want to implement it for VMs, as a second phase.

> Furthermore, I'm very confused about the chosen points. We had all of this (except items 12 and 13) in a PR months ago.

The code base does not look the same as it was months ago (roughly two..), rebasing is not an option IMO, we can take few parts from the PoC and adapt it.

> * the MVP for unplug functionality would be composed of steps 6 through 8, inclusive.

Reviewer: My point is we know now what we did 2 months ago. Nothing changed in regard to this. Did it ? 

The motivation for removing unplug from the original PR was we would leak both the networking infrastructure, and the dhcp server on interface unplug. Thus we faced a choice: do we want to provide a full feature (and for that have to bloat the PR some more by adding the code to prevent the leaks), or do we want to have half the feature (only hotplug), but a smaller PR ? We chose the latter.

From what I see here, we'll move forward without having addressed any of that - IMO, that's concerning. 

Wouldn't it make more sense to address the gaps first ? 

If the leaks don't matter because we can use the feature (without leaks) to plug interfaces if we use VM migration, that information is important, and, AFAIU, it not stated anywhere. Please add it.

> * the MVP for unplug functionality would be composed of steps 6 through 8, inclusive.

Author: This is a mere work division suggestion, the feature shall include all the functionalities we had in mind two months ago, including the cleanup.
Hopefully, dividing it into smaller steps will get us to the target faster. At least make the review process easier.

> If the leaks don't matter because we can use the feature (without leaks) to plug interfaces if we use VM migration, that information is important, and, AFAIU, it not stated anywhere. Please add it.

Thanks for the heads up, I will add it.

> * the unplug MVP does not handler.

Author: This is a leftover sorry for the noise.
